### DISCUSSION TITLE
Dnp3,Plc4x and IEC61850

### DISCUSSION THREAD

**Discussion by rudyroc** [riclolsen/json-scada#150] [General]
Dear Ricardo
After a gap year, I'm here again.
I have tested version 37 of Json-Scada running under Linux using the Rocky 9 distro. Also I have tested the same version using docker on Alpine Linux 18 distro.

DNP3 Client comments:
After some small modifications the DNP3 client works correctly. I attach a file with the modifications to the docker-compose.yaml files, 
both compile-docker and demo-docker. I also attach the supervisor dnp3_client.ini file used whith Rocky Linux 9. Wine version installed is the latest 64-bit one.

PLC4X Client comments: 
When trying Modbus Serial this didn't work. I attach screenshot of the logs, both in Linux and Windows. One question, are the endpointURLs used correct?

IEC61850 Client comments:
In the log file, I can see what the switch position is retrived as double bit ( 0=undefined/1=open/2=closed/3=error), but in the database are recorded as single bit (0 = open / 1 = closed).
Any suggestion?

Best regards,
Rudy
[json-scada_dnp3_docker_modifications.txt](https://github.com/user-attachments/files/17004616/json-scada_dnp3_docker_modifications.txt)
![log_plcx4client_serial_modbus_linux](https://github.com/user-attachments/assets/2ac8b17f-215d-4a36-89ad-c12da5fa8ddd)
![log_plcx4client_serial_modbus_windows](https://github.com/user-attachments/assets/09aea07d-dc57-470e-abaf-28b2a56cc528)
[rocky_linux_9_supervisor_dnp3_client.ini.txt](https://github.com/user-attachments/files/17004617/rocky_linux_9_supervisor_dnp3_client.ini.txt)

**Comment by riclolsen** [comment]
Welcome back Rudy!
After some struggling with wine I was able to make DNP3 and OPC-DA work on Linux. It was necessary revert to Dotnet6.
You can test the new compilation and demo script. Thanks for the contribution.

**Comment by riclolsen** [comment]
PLC4X: 
I only tested Modbus TCP. I am not sure PLC4GO works for Modbus serial. 
I will look into that eventually.

**Comment by riclolsen** [comment]
IEC61850 
In realtimeData collection the digital value has 1 or 0 values, this is by design.
However the driver should manage the failed and transit flags to reflect the 0 and 3 states, it is not setting the transit flag yet, I will fix that.

**Reply by riclolsen** [reply]
The code was updated to handle the undefined state. Please test if possible, I don't have an easy way to test this situation.

**Reply by rudyroc** [reply]
Hi Ricardo !
Thank you for your response. I will try the modifications to IEC61850 and the new version of the Docker as soon as possible. 
I will keep you updated on the results.
Rudy

**Comment by rudyroc** [comment]
Hi Ricardo 
I compiled and tested the new version of the 61850 driver (0.1.18). The value of the "transient" flag in the database is always "false", even when the value received by the driver is 0 or 3.
Best regards,
Rudy
